Implementation of DB-API for BigQuery.

tswast opened the pull request:
Implements `Cursor.execute()` and `Cursor.fetchone()` without support
for query parameters.

Tested manually with a Jupyter notebook
```
# In[1]:
from google.cloud import bigquery
from google.cloud.bigquery import bqdb
connection = bqdb.connect()
cursor = connection.cursor()

# In[2]:
cursor.execute("SELECT (1 + 2) AS s;")

# In[3]:
cursor.fetchone()

# In[4]:
cursor.fetchone()

# In[5]:
cursor.description

# In[6]:
cursor.rowcount

# In[7]:
cursor.execute("DELETE FROM `swast-scratch.hello_world.hello` WHERE id = 1;")

# In[8]:
cursor.rowcount
```

Makes progress on https://github.com/GoogleCloudPlatform/google-cloud-python/issues/2434

jonparrott commented on bigquery/google/cloud/bigquery/bqdb/__init__.py:
```diff
@@ -0,0 +1,49 @@
+
+apilevel = "2.0"
+
+# Threads may share the module, but not connections.
```
I'm guessing if we switch to a threadsafe transport then we can use `threadsafety = 2`?

tswast replied:
I made my best guess. The DB-API has a bigquery.Client, so thread safety should be the same as we have for that. We don't talk about thread safety at all in https://googlecloudplatform.github.io/google-cloud-python/stable/bigquery-usage.html

jonparrott replied:
This is correct, I was just making sure that the rest of this design didn't do anything that would prevent us from bumping the threadsafety up once the client is threadsafe.

jonparrott commented on bigquery/google/cloud/bigquery/bqdb/connection.py:
```diff
@@ -0,0 +1,55 @@
+
+
+class Connection(object):
+    """Connection to Google BigQuery.
```
I'd prefix all of these with `DB-API`, e.g. `DB-API Connection ...`, just for the sake of people who might search and land here.

jonparrott commented on bigquery/google/cloud/bigquery/bqdb/cursor.py:
```diff
@@ -0,0 +1,102 @@
+            return None
+
+        rows, _, page_token = self._query_results.fetch_data(
+                max_results=1, page_token=self._page_token)
```
Weird indent?

jonparrott commented on bigquery/google/cloud/bigquery/bqdb/cursor.py:
```diff
@@ -0,0 +1,102 @@
+        #       infer types from parameter inputs.
+        query_job = client.run_async_query(job_id, operation)
+        query_job.use_legacy_sql = False
+        query_job.begin()
```
meta: are This could use some blank newlines to separate the logic for readability, so could several of the tests.

jonparrott commented on bigquery/google/cloud/bigquery/bqdb/cursor.py:
```diff
@@ -0,0 +1,102 @@
+        self._has_fetched_all_rows = False
+        client = self.connection._client
+        job_id = str(uuid.uuid4())
+        # TODO: parameters: if not ``None``, check if ``dict`` or sequence and
```
Do you plan to resolve this in this PR?

tswast replied:
Should I? Query parameter support is going to be a lot more code, so I was waiting to do it in a separate PR.

jonparrott replied:
I'd prefer not to merge incomplete implementations here, as it might block us from release or make us release partially implemented features. @dhermes might feel differently, and if he does, then I defer to him.

jonparrott commented on bigquery/google/cloud/bigquery/bqdb/cursor.py:
```diff
@@ -0,0 +1,102 @@
+        if self._has_fetched_all_rows:
+            return None
+
+        rows, _, page_token = self._query_results.fetch_data(
```
Does bigquery not implement our new iterator protocol? @dhermes?

tswast replied:
Nope. https://github.com/GoogleCloudPlatform/google-cloud-python/issues/2840

jonparrott replied:
I would be nice to have that before we go too deep down the rabbit hole here.

jonparrott commented on bigquery/google/cloud/bigquery/bqdb/__init__.py:
```diff
@@ -0,0 +1,49 @@
+# Copyright 2016 Google Inc.
```
I'd prefer the import path to be `google.cloud.bigquery.dbapi`.

Bamieh commented:
thanks for the effort! i want to use big query with superset via sqlalchemy, i believe this is a good starting point, im willing to help out if needed!

tseaver commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,102 @@
+            self._has_fetched_all_rows = True
+
+        self._page_token = page_token
+        return rows[0]
```
Needs to stash any remaining rows, plus the page token, and use them on subsequent calls.  Also, needs to raise `Error` if `executeone()` failed or was never called.

tseaver replied:
Also, seems like `fetchmany` and `fetchall` should be implemented, even in the first pass.

tseaver commented on bigquery/tests/system.py:
```diff
@@ -775,6 +778,13 @@ def test_sync_query_w_query_params(self):
+                self.assertEqual(len(row), 1)
+                self.assertEqual(row[0], example['expected'])
+                row = Config.CURSOR.fetchone()
+                self.assertIsNone(row)
```
Needs to have the `dbapi` tests separated out into their own methods (factoring out the generation of `EXAMPLES`.

tseaver commented on bigquery/tests/unit/test_dbapi_connection.py:
```diff
@@ -0,0 +1,73 @@
+        from google.cloud.bigquery import Client
+        from google.cloud.bigquery.dbapi import connect
+        from google.cloud.bigquery.dbapi import Connection
+        connection = connect()
```
This test needs to mock out the environment-sniffing such that it runs without any of the "normal" environment variables set.

tseaver commented on bigquery/tests/unit/test_dbapi_cursor.py:
```diff
@@ -0,0 +1,134 @@
+        from google.cloud.bigquery.dbapi import Cursor
+        connection = connect(_Client())
+        cursor = connection.cursor()
+        row = cursor.fetchone()
```
[PEP 249](https://www.python.org/dev/peps/pep-0249/#fetchone) says this case should raise an `Error`.

tswast commented:
Hold off on reviewing. I still need to address a few things from your last review & implement query parameters.

tswast commented:
I've pushed a new commit. Should be ready to review. (I'll be making a couple extra unit tests to make coverage report happy, but with the integration tests, I'm pretty confident this is working.)

I believe I've addressed most of your comments. I'll file issues for the TODOs once we're confident the PR won't change much before merging.

tswast commented:
Coverage back @ 100%. @jonparrott PTAL

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/__init__.py:
```diff
@@ -0,0 +1,61 @@
+   or deprecation policy.
+"""
+
+from google.cloud.bigquery.dbapi.connection import connect  # noqa
```
Don't `noqa`, these, please use a `__all__` statement. https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/__init__.py:
```diff
@@ -0,0 +1,61 @@
+apilevel = "2.0"
+
+# Threads may share the module, but not connections.
+threadsafety = 1
```
Can you file a tracking bug to bump this up to 2 once we dump httplib2? #1998 related.

dhermes replied:
`1` means `False`, or its some enum?

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,138 @@
+def scalar_to_query_parameter(name=None, value=None):
+    """Convert a scalar value into a query parameter.
+
+    Note: the bytes type cannot be distinguished from a string in Python 2.
```
This comment is misleading. There is always a text and bytes type, it's just that in Python 2 the default string syntax yields bytes. You could enforce strict separation by *requiring unicode* by checking for `six.text_type` or `STRING` and using `six.binary_type` for `BYTES`.

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,138 @@
+    for value in parameters:
+        query_parameters.append(scalar_to_query_parameter(value=value))
+
+    return query_parameters
```
Perhaps you could express this as:

```python
return [scalar_to_query_parameter(value) for value in parameters]
```

tswast replied:
(I think I didn't do this to begin with because I wanted to do something to support struct/list parameters, but I'd rather wait on that is it's a pretty unique-to-BigQuery feature. I wouldn't expect many people to try to pass in a dict as a singular parameter. Filed https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3524

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,138 @@
+        value = parameters[name]
+        query_parameters.append(scalar_to_query_parameter(name, value))
+
+    return query_parameters
```
perhaps you could express this as:

```python
return [scalar_to_query_parameter(name, value) for name, value in six.iteritems(parameters)]
```

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,138 @@
+    if parameters is None:
+        return []
+
+    if isinstance(parameters, collections.Mapping):
```
thank you for using `collections.Mapping` :)

tswast replied:
:+1:

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,138 @@
+def to_query_parameters(parameters):
+    """Converts DB-API parameter values into query parameters.
+
+    :type: dict or list
```
`Mapping[str, Any] or Sequence[Any]`

tswast replied:
Done

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,261 @@
+from google.cloud.bigquery.dbapi import exceptions
+
+
+_ARRAYSIZE_DEFAULT = 20
```
Docstring/comment explaining this constant's purpose?

tswast replied:
Removed this constant (it was incorrect, anyway). I re-read the PEP and the arraysize is supposed to default to 1.

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,261 @@
+    def __init__(self, connection):
+        self.connection = connection
+        self.description = None
+        self.rowcount = -1
```
Can you add a comment here explaining "Per PEP 239: The attribute is -1 in case no .execute*() has been performed on the cursor or the rowcount of the last operation is cannot be determined by the interface."

tswast replied:
Done.

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,261 @@
+
+        desc = []
+        for field in schema:
+            desc.append(tuple([
```
Can you use a NamedTuple that matches the definition in PEP 249? https://www.python.org/dev/peps/pep-0249/#description

tswast replied:
Done.

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,261 @@
+        self.rowcount = total_rows
+
+    def _format_operation_list(self, operation, parameters):
+        """Formats parameters in operation in way BigQuery expects.
```
Can you define here what you mean by `in the way bigquery expects` can you show an example transformation?

tswast replied:
Added an example.

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,261 @@
+            raise exceptions.ProgrammingError(ex)
+
+    def _format_operation_dict(self, operation, parameters):
+        """Formats parameters in operation in way BigQuery expects.
```
Same here.

tswast replied:
Done.

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/types.py:
```diff
@@ -0,0 +1,75 @@
+Timestamp = datetime.datetime
+DateFromTicks = datetime.date.fromtimestamp
+TimestampFromTicks = datetime.datetime.fromtimestamp
+Binary = bytes
```
`six.binary_type` for posterity.

tswast replied:
Done.

jonparrott commented on bigquery/tests/system.py:
```diff
@@ -819,6 +901,93 @@ def test_sync_query_w_query_params(self):
             self.assertEqual(query.rows[0][0], example['expected'])
 
+    def test_dbapi_w_query_parameters(self):
+        EXAMPLES = [
```
this is not a constant, it should be `lower_snake_case`.

tswast replied:
done.

jonparrott commented on bigquery/tests/unit/test_dbapi__helpers.py:
```diff
@@ -0,0 +1,146 @@
+            self.assertEqual(named_parameter.type_, expected_type, msg=msg)
+            self.assertEqual(named_parameter.value, value, msg=msg)
+
+    @unittest.skipIf(six.PY2, 'Bytes cannot be distinguished from string.')
```
I would strongly prefer you to enforce unicode usage on Python 2.

tswast replied:
Done.

tswast commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,33 @@
+        if job.state == 'DONE':
+            if job.error_result:
+                # TODO: raise a more specific exception, based on the error.
+                # See: https://cloud.google.com/bigquery/troubleshooting-errors
```
Filed https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3523

jonparrott commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+            return
+
+        self.description = tuple([
+            Column(
```
Nit: Can you use named arguments to set the value here so it's clear what you're setting to None?

jonparrott commented:
@tseaver @dhermes This LGTM, can one of you do a final pass?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/__init__.py:
```diff
@@ -0,0 +1,70 @@
+# Copyright 2016 Google Inc.
```
2017

dhermes commented on bigquery/google/cloud/bigquery/dbapi/__init__.py:
```diff
@@ -0,0 +1,70 @@
+from google.cloud.bigquery.dbapi.types import STRING
+
+
+apilevel = "2.0"
```
Can you use single quotes?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/__init__.py:
```diff
@@ -0,0 +1,70 @@
+# Threads may share the module, but not connections.
+threadsafety = 1
+
+paramstyle = "pyformat"
```
Ditto here, single quotes.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+# Copyright 2016 Google Inc.
```
2017

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+import collections
+import datetime
+import numbers
+import six
```
`six` is not a standard library important, please put it in it's own 3rd party section

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+def wait_for_job(job):
+    """Waits for a job to complete by polling until the state is `DONE`.
+
+    Raises a DatabaseError if the job fails.
```
1. Can you convert this to a `:raises:` section?
1. You should mention the `sleep()`

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+
+    Raises a DatabaseError if the job fails.
+
+    :type: :class:`~google.cloud.bigquery.job._AsyncJob`
```
```
:type job: ...
```

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+def scalar_to_query_parameter(value, name=None):
+    """Convert a scalar value into a query parameter.
+
+    Note: You must use the unicode type for string parameters in Python 2.
```
You could use the `.. note::` directive in Sphinx? Also, you might just copy-paste the [explanation][1] we have in datastore.

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/cbf073bc5469bac9398a5bf7aa5e93e6cb637211/datastore/google/cloud/datastore/entity.py#L111-L119

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+
+    Note: You must use the unicode type for string parameters in Python 2.
+
+    Raises a :class:`~ google.cloud.bigquery.dbapi.exceptions.ProgrammingError`
```
1. Please convert this to a `:raises:` section
1. You have a space between `~` and `google` that doesn't belong

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+    For more information about BigQuery data types, see:
+    https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types
+
+    :type: any
```
`:type value:`

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+    :type: any
+    :param value: A scalar value to convert into a query parameter.
+
+    :type: str
```
`:type name:`

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+    :type: str
+    :param name: Optional name of the query parameter.
+
+    :rtype: :class:`~google.cloud.bigquery.ScalarQueryParameter`
```
Please also add a `:returns:` explain what is returned

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+def to_query_parameters_list(parameters):
+    """Converts a list of parameter values into query parameters.
+
+    :type: Sequence[Any]
```
`type parameters`

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+    :param parameters: Sequence of query parameter values.
+
+    :rtype:
+        list of :class:`~google.cloud.bigquery._helpers.AbstractQueryParameter`
```
If you are going in the `Sequence[Any]` territory you could do `List[foo, ...]` here. Though it won't render in a nice way until #3518 is in.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+def to_query_parameters_dict(parameters):
+    """Converts a dictionary of parameter values into query parameters.
+
+    :type: Mapping[str, Any]
```
`:type parameters:`

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+    return [
+        scalar_to_query_parameter(value, name=name)
+        for name, value
+        in six.iteritems(parameters)]
```
Order is not important here? Worth mentioning in the docstring?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/connection.py:
```diff
@@ -0,0 +1,56 @@
+# Copyright 2016 Google Inc.
```
2017

dhermes commented on bigquery/google/cloud/bigquery/dbapi/_helpers.py:
```diff
@@ -0,0 +1,131 @@
+    elif isinstance(value, six.binary_type):
+        parameter_type = 'BYTES'
+    elif isinstance(value, datetime.datetime):
+        parameter_type = 'TIMESTAMP' if value.tzinfo else 'DATETIME'
```
1. Please use an explicit [`None` check][1]
1. Help me with the distinction here, why does timezone info make it a datetime vs. a timestamp?

[1]: http://iruntheinternet.com/lulzdump/images/nunchucks-nuns-chuck-norris-nun-12593488599.jpg?id=

tswast replied:
Re 2:

In BigQuery a [datetime](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#datetime-type) is a civil time (no timezone) (example: Global Pi Day Celebration March 14, 2017 at 1:59pm) whereas a [timestamp](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp-type) represents an absolute point in time (including timezone, example: September 4, 1998 at 9am, Los Angeles timezone)

dhermes replied:
OK cool thanks for the clarification. We should definitely document that on a public class / method.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+                internal_size=None,
+                precision=None,
+                scale=None,
+                null_ok=field.mode == 'NULLABLE')
```
Worth adding ` field.is_nullable()` method?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+
+        try:
+            return operation % tuple(formatted_params)
+        except TypeError as ex:
```
Pylint is currently turned off but it is coming back, and it will yell at you for a 2-char variable name. Might I suggest `exc`?

tswast replied:
done.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+            total_rows = num_dml_affected_rows
+        self.rowcount = total_rows
+
+    def _format_operation_list(self, operation, parameters):
```
This doesn't use `self`, it should be a `@staticmethod` or just make it a function in the module

tswast replied:
Done.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        :type: Sequence[Any]
+        :param parameters: Sequence of parameter values.
+        """
+        formatted_params = ['?' for _ in parameters]
```
Why would anyone want that?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        """
+        formatted_params = {}
+        for name in parameters:
+            formatted_params[name] = '@{}'.format(name)
```
What if the name isn't simple, i.e. has chars like `{}` or quotes or hyphens?

tswast replied:
Then BigQuery will return a syntax error and the job will fail.

I'll check with the eng team if there's a way I can escape characters.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        try:
+            return operation % formatted_params
+        except KeyError as ex:
+            raise exceptions.ProgrammingError(ex)
```
`ex` -> `exc`

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        except TypeError as ex:
+            raise exceptions.ProgrammingError(ex)
+
+    def _format_operation_dict(self, operation, parameters):
```
`@staticmethod` or function

tswast replied:
done.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        if isinstance(parameters, collections.Mapping):
+            return self._format_operation_dict(operation, parameters)
+
+        return self._format_operation_list(operation, parameters)
```
This would be an argument for those helpers to remain as `@staticmethod`s

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        self._page_token = None
+        self._has_fetched_all_rows = False
+        client = self.connection._client
+        job_id = str(uuid.uuid4())
```
Is it expected that the client determines the job ID?

lukesneeringer replied:
The BigQuery team had separately asked for this.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        """
+        self._query_results = None
+        self._page_token = None
+        self._has_fetched_all_rows = False
```
1. This is like a `reset()`?
1. This means you expect a single `Cursor` instance to be `execute()`-d many times?
1. Worth having a `reset() / _reset()` helper?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+            query_parameters=query_parameters)
+        query_job.use_legacy_sql = False
+        query_job.begin()
+        _helpers.wait_for_job(query_job)
```
1. Why not just use a sync query if we're going to wait anyhow?
1. Is it worth making this helper just part of the job as a method?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+            pass
+
+        self._query_data = iter(
+            query_results.fetch_data(max_results=self.arraysize))
```
1. Can't you use this to set `total_rows`?
1. Is it not already an iterator? (i.e. why call `iter()`?)

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        self._set_rowcount(query_results)
+        self._set_description(query_results.schema)
+
+    def executemany(self, operation, seq_of_parameters):
```
Ick. Bad PEP 249, bad! Worth offering an alias `execute_many` that actually distinguishes word boundary?

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        :rtype: tuple
+        :returns:
+            A tuple representing a row or ``None`` if no more data is
+            available.
```
Missing a `:raises:` section

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+                'No query results: execute() must be called before fetch.')
+
+        try:
+            return six.next(self._query_data)
```
So this might be a fetch, or it might be just grabbing an item from a page (though the default page size is 1?)

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+            ``arraysize`` property value.
+
+        :rtype: Sequence[tuple]
+        :returns: A list of rows.
```
Missing a `:raises:` section

dhermes commented on bigquery/google/cloud/bigquery/dbapi/cursor.py:
```diff
@@ -0,0 +1,280 @@
+        """Fetch all remaining results from the last ``execute*()`` call.
+
+        :rtype: Sequence[tuple]
+        :returns: A list of all the rows in the results.
```
Missing `:raises:` section

dhermes commented on bigquery/google/cloud/bigquery/dbapi/exceptions.py:
```diff
@@ -0,0 +1,58 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Exceptions used in the Google BigQuery DB-API."""
```
This module is so WAT to me. @jonparrott @tseaver @lukesneeringer is it common for a PEP to be so opinionated about errors like this?

----

Directly from the PEP:

> This is the exception inheritance layout:
> 
> ```
> StandardError
> |__Warning
> |__Error
>   |__InterfaceError
>    |__DatabaseError
>       |__DataError
>       |__OperationalError
>       |__IntegrityError
>       |__InternalError
>       |__ProgrammingError
>       |__NotSupportedError
> ```

dhermes commented on bigquery/google/cloud/bigquery/dbapi/types.py:
```diff
@@ -0,0 +1,77 @@
+
+See `PEP 249`_ for details.
+
+.. _`PEP 249`:
```
You don't want / need the backticks here? (Though maybe Sphinx is cool with it, I'm not sure.)

dhermes commented on bigquery/google/cloud/bigquery/dbapi/types.py:
```diff
@@ -0,0 +1,77 @@
+Timestamp = datetime.datetime
+DateFromTicks = datetime.date.fromtimestamp
+TimestampFromTicks = datetime.datetime.fromtimestamp
+Binary = six.binary_type
```
This will work fine for `isinstance` checks but `bytes(u'a')` fails in Python 3, so I'm not sure what the constructor must be? The PEP says

```
Binary ( string )
```

This function constructs an object capable of holding a binary (long) string value.

dhermes commented on bigquery/google/cloud/bigquery/dbapi/types.py:
```diff
@@ -0,0 +1,77 @@
+    def __eq__(self, other):
+        if other in self.values:
+            return True
+        return False
```
You can just `return other in self.values`

dhermes commented on bigquery/tests/system.py:
```diff
@@ -70,10 +71,12 @@ class Config(object):
 
 def setUpModule():
     Config.CLIENT = bigquery.Client()
+    Config.CURSOR = dbapi.connect().cursor()
```
Shouldn't you pass `Config.CLIENT` to `connect()`?

dhermes commented on bigquery/tests/system.py:
```diff
@@ -167,9 +170,9 @@ def test_list_datasets(self):
-            self.to_delete.append(dataset)
+            created_dataset = Config.CLIENT.dataset(dataset_name)
+            retry_403(created_dataset.create)()
+            self.to_delete.append(created_dataset)
```
This could be in it's own PR? It's just hygiene?

tswast replied:
Filed the cleanup to its own PR: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/3551

dhermes commented on bigquery/tests/system.py:
```diff
@@ -222,9 +225,9 @@ def test_list_tables(self):
-            self.to_delete.insert(0, table)
+            created_table = dataset.table(table_name, schema=[full_name, age])
+            created_table.create()
+            self.to_delete.insert(0, created_table)
```
This could be in it's own PR? It's just hygiene?

dhermes commented on bigquery/tests/system.py:
```diff
@@ -544,7 +547,7 @@ def test_sync_query_w_legacy_sql_types(self):
         stamp = '%s %s' % (naive.date().isoformat(), naive.time().isoformat())
         zoned = naive.replace(tzinfo=UTC)
-        EXAMPLES = [
+        examples = [
```
:grin: Thank you

dhermes commented on bigquery/tests/system.py:
```diff
@@ -659,14 +662,93 @@ def test_sync_query_w_standard_sql_types(self):
+            self.assertEqual(Config.CURSOR.rowcount, 1)
+            row = Config.CURSOR.fetchone()
+            self.assertEqual(len(row), 1)
+            self.assertEqual(row[0], example['expected'])
```
You could also just:

```python
self.assertEqual(row, [example['expected']])
```

lukesneeringer replied:
I prefer the current form.

dhermes commented on bigquery/tests/system.py:
```diff
@@ -659,14 +662,93 @@ def test_sync_query_w_standard_sql_types(self):
+                )
+
+        def _job_done(instance):
+            return instance.state.lower() == 'done'
```
Can you define this at class-level or module level? (I am not a big fan of "on-the-fly" functions, mostly because they can become closures on accident.)

dhermes commented on bigquery/tests/system.py:
```diff
@@ -659,14 +662,93 @@ def test_sync_query_w_standard_sql_types(self):
+            'UPDATE {}.{} '
+            'SET greeting = \'Guten Tag\' '
+            'WHERE greeting = \'Hello World\''.format(
+                dataset_name, table_name))
```
This is a lot to process. Can you set up the template separately from the formatting? Also, feel free to use double quotes here to make the text easier to read (your call though).

dhermes commented on bigquery/tests/system.py:
```diff
@@ -659,14 +662,93 @@ def test_sync_query_w_standard_sql_types(self):
+        with _NamedTemporaryFile() as temp:
+            with open(temp.name, 'w') as csv_write:
+                writer = csv.writer(csv_write)
+                writer.writerow(('Greeting'))
```
This is not a `tuple`, you need a trailing comma.

dhermes commented on bigquery/tests/system.py:
```diff
@@ -819,6 +901,95 @@ def test_sync_query_w_query_params(self):
+                Config.CURSOR.execute(
+                    example['sql'], example['query_parameters'])
+            except dbapi.DatabaseError as ex:
+                raise dbapi.DatabaseError('{} {}'.format(ex, msg))
```
1. `ex` -> `exc`
1. Don't `raise` in a test, just fail the test.

dhermes commented on bigquery/tests/system.py:
```diff
@@ -819,6 +901,95 @@ def test_sync_query_w_query_params(self):
+            self.assertEqual(len(row), 1, msg=msg)
+            self.assertEqual(row[0], example['expected'], msg=msg)
+            row = Config.CURSOR.fetchone()
+            self.assertIsNone(row, msg=msg)
```
`msg=msg` is just for printing on failure?

tswast replied:
Yeah, I added that because I was getting some failures and couldn't tell which example was failing from the message. Seemed like it was good to keep in there.

dhermes commented on bigquery/tests/system.py:
```diff
@@ -838,7 +1009,6 @@ def test_large_query_w_public_data(self):
         SQL = 'SELECT * from `{}.{}.{}` LIMIT {}'.format(
             PUBLIC, DATASET_NAME, TABLE_NAME, LIMIT)
 
-        dataset = Config.CLIENT.dataset(DATASET_NAME, project=PUBLIC)
```
Unused variable? Thanks

tswast commented:
Oops, I think GitHub sent my review early. I haven't uploaded my fixes yet.

tswast commented:
Okay. I just pushed my latest changes.

tswast commented:
Ready for another review pass when you get a chance.

tswast commented:
I've rebased on the latest master. Okay to merge?
